// eslint-plugin-yaml/dist/index.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const js_yaml_1 = require("js-yaml");
const jshint_1 = require("jshint");
const package_json_1 = __importDefault(require("../package.json"));
const path_1 = __importDefault(require("path"));
function isYaml(fileName) {
    const fileExtension = path_1.default.extname(fileName);
    return [".yaml", ".yml"].includes(fileExtension);
}
const fileContents = new Map();
function preprocess(text, fileName) {
    if (!isYaml(fileName)) {
        return [];
    }
    fileContents.set(fileName, text);
    return [{ text, filename: fileName }];
}
function postprocess(_messages, fileName) {
    var _a, _b;
    if (!isYaml(fileName)) {
        return [];
    }
    let linter_messages = [];
    const fileContent = fileContents.get(fileName);
    if (fileContent !== undefined) {
        let yamlDocs;
        try {
            yamlDocs = loadYaml(fileContent, fileName);
        }
        catch (e) {
            const { message, mark } = e;
            return [
                {
                    ruleId: "invalid-yaml",
                    severity: 2,
                    message,
                    source: mark === null || mark === void 0 ? void 0 : mark.buffer,
                    line: (_a = mark === null || mark === void 0 ? void 0 : mark.line) !== null && _a !== void 0 ? _a : 0,
                    column: (_b = mark === null || mark === void 0 ? void 0 : mark.column) !== null && _b !== void 0 ? _b : 0,
                },
            ];
        }
        const errors = yamlDocs.flatMap((yamlDoc) => lintJSON(yamlDoc));
        linter_messages = errors.map((error) => {
            const { reason, evidence, line, character } = error;
            return {
                ruleId: "bad-yaml",
                severity: 2,
                message: reason,
                source: evidence,
                line,
                column: character,
            };
        });
        fileContents.delete(fileName);
    }
    return linter_messages;
}
function loadYaml(fileContent, fileName) {
    return (0, js_yaml_1.loadAll)(fileContent, undefined, {
        filename: fileName,
        json: false,
    });
}
function lintJSON(yamlDoc) {
    var _a;
    const yaml_json = JSON.stringify(yamlDoc, null, 2);
    (0, jshint_1.JSHINT)(yaml_json);
    const data = jshint_1.JSHINT.data();
    const errors = (_a = data === null || data === void 0 ? void 0 : data.errors) !== null && _a !== void 0 ? _a : [];
    return errors;
}
const processors = {
    [package_json_1.default.name]: {
        meta: {
            name: package_json_1.default.name,
            version: package_json_1.default.version,
        },
        preprocess,
        postprocess,
    },
};
const meta = {
    name: package_json_1.default.name,
    version: package_json_1.default.version,
    type: "problem",
    docs: {
        description: "YAML linting",
        category: "Parsing Issues",
        recommended: false,
        url: "https://github.com/aminya/eslint-plugin-yaml",
    },
    fixable: "code",
    schema: [],
};
const plugin = {
    meta,
    processors,
    configs: {
        recommended: {},
        legacy: {},
    },
};
const recommendedConfig = {
    name: `${package_json_1.default.name}/recommended}`,
    files: ["**/*.yaml", "**/*.yml", "!**/node_modules/**", "!**/pnpm-lock.yaml", "**/.github/**.{yml,yaml}"],
    processor: {
        name: package_json_1.default.name,
        preprocess,
        postprocess,
    },
    plugins: {
        [package_json_1.default.name]: plugin,
    },
};
plugin.configs.recommended = recommendedConfig;
const legacyConfig = {
    overrides: [
        {
            plugins: [package_json_1.default.name],
            files: ["**/*.yml", "**/*.yaml", "!**/node_modules/**", "!**/pnpm-lock.yaml", "**/.github/**.{yml,yaml}"],
            processor: `yaml/${package_json_1.default.name}`,
        },
    ],
};
plugin.configs.legacy = legacyConfig;
exports.default = plugin;
module.exports = plugin;
module.exports.default = plugin;
